src/lazy/binary/raw/v1_1/value.rs

```diff
@@ -265,9 +268,298 @@ impl<'top> LazyRawBinaryValue_1_1<'top> {
+        let min = (value_bytes[2] & 0xE0) >> 5; // 3 LSbits
+        let min = min | ((value_bytes[3] & 0x07) << 3); // 3 MSbits
+        let ts_builder = ts_builder.with_hour_and_minute(hour as u32, min as u32);
+
+        // We're at least Hour&Minute so we need either an offset, or indicator that the timestamp
+        // is UTC or Unknown.
+
+        // UTC / Unkonwn Offset
+        if length_code < 8 {
+            self.read_timestamp_short_no_offset(value_bytes, ts_builder)
+        } else {
+            // Known Offset
+            self.read_timestamp_short_offset(value_bytes, ts_builder)
+        }
+    }
+
+    // Helper method callsed by [`Self::read_timestamp_short`]. Reads the time information from a
+    // timestamp encoded in long form.
+    fn read_timestamp_long(&self) -> ValueParseResult<'top, BinaryEncoding_1_1> {
+        use crate::lazy::encoder::binary::v1_1::fixed_uint::FixedUInt;
+        use crate::lazy::encoder::binary::v1_1::flex_uint::FlexUInt;
+        use crate::types::decimal::*;
+
+        let value_bytes = self.value_body()?;
+        let value_length = self.encoded_value.value_body_length;
+
+        if value_length < 2 || value_length == 4 || value_length == 5 {
+            return Err(IonError::decoding_error("invalid timestamp length"));
+        }
+
+        let year = (value_bytes[0] as u32) | ((value_bytes[1] & 0x3F) as u32) << 8;
+        let ts_builder = Timestamp::with_year(year);
+        if value_length == 2 {
+            return Ok(RawValueRef::Timestamp(ts_builder.build()?));
+        }
+
+        let month = ((value_bytes[1] & 0xC0) as u32) >> 6 | ((value_bytes[2] & 0x03) as u32) << 2;
+        let day = ((value_bytes[2] & 0x7C) as u32) >> 2;
+        let ts_builder = ts_builder.with_month(month);
+        if value_length == 3 && day == 0 {
+            return Ok(RawValueRef::Timestamp(ts_builder.build()?));
+        }
+
+        let ts_builder = ts_builder.with_day(day);
+        if value_length == 3 {
+            return Ok(RawValueRef::Timestamp(ts_builder.build()?));
+        }
+
+        let hour = ((value_bytes[2] & 0x80) as u32) >> 7 | ((value_bytes[3] & 0x0F) as u32) << 1;
+        let minute = ((value_bytes[3] & 0xF0) as u32) >> 4 | ((value_bytes[4] & 0x03) as u32) << 4;
+        let offset = ((value_bytes[4] & 0xFC) as u32) >> 2 | ((value_bytes[5] & 0x3F) as u32) << 6;
+        let offset = if (offset & 0x800) == 0x800 && offset != 0xFFF {
+            (offset | !0xFFFu32) as i32 // Extend sign for negative 12-bit offsets.
+        } else {
+            offset as i32
+        };
+        let ts_builder = ts_builder.with_hour_and_minute(hour, minute);
+
+        if value_length == 6 {
+            if offset != 0x0FFF {
+                let ts_builder = ts_builder.with_offset(offset);
+                return Ok(RawValueRef::Timestamp(ts_builder.build()?));
+            }
+            return Ok(RawValueRef::Timestamp(ts_builder.build()?));
```

Comment:
You can get rid of the double `return`s like this:
```suggestion
            let ts_builder = if offset == 0x0FFF { ts_builder } else {
                ts_builder.with_offset(offset);
            }
            return Ok(RawValueRef::Timestamp(ts_builder.build()?));
```

If you take my previous suggestion to use `Option<i32>` for the offset, then it can be further simplified to
```rust
let ts_builder = offset.map_or(ts_builder, |o| ts_builder.with_offset(o));
return Ok(RawValueRef::Timestamp(ts_builder.build()?));
```

...which, I suppose, is a signal that the `TimestampBuilder.with_offset` should probably be changed to accept `Option<i32>`.

Reply:
Unfortunately, TimestampBuilder's precision is managed through a generic parameter, so each precision change also changes the type. So the if-else fails to compile because its branches produce 2 different types. That's the bulk of the reason for carrying the offset and setting it only when the timestamp is built.

Reply:
Oh, right. I forgot about that.